File: StatePattern/StateDemo.cs
enum State {
  open,
  closed,
  error,
}

class Connection {

  State state;

  public Connection() {
    state = State.closed;
  }

  public void Connect() {
    if (state == State.open)
      Console.WriteLine("Connection already open");
    else if (state == State.error)
      Console.WriteLine("Connection is error state");
    state = State.open;
  }

  public void SendMessage(String? message) {
    if (message == null)
      state = State.error;

    if (state == State.error)
      Console.WriteLine("Connection is in error state");
    else if (state == State.open)
      Console.WriteLine("Sending message " + message);
    else
      Console.WriteLine("Connection is not open");
  }

  public void Close() {
    if (state == State.closed)
      Console.WriteLine("Connection is not open");
  }
}

public class StateDemo {
  public static void Main() {
    var connection = new Connection();
    connection.SendMessage("Hello");
    connection.Connect();
    connection.SendMessage("Hello");
    connection.Connect();
    connection.SendMessage(null);
    connection.Close();
  }
}

File: StatePattern/StateDemo.py
from enum import Enum


class State(Enum):
    open = 1
    closed = 2
    error = 3


class Connection:
    def __init__(self) -> None:
        self.state: State = State.closed

    def connect(self) -> None:
        if self.state == State.open:
            print("Connection already open")
        elif self.state == State.error:
            print("Connection is error state")
        self.state = State.open

    def send_message(self, message: str | None) -> None:
        if message is None:
            self.state = State.error

        if self.state == State.error:
            print("Connection is in error state")
        elif self.state == State.open:
            print(f"Sending message {message}")
        else:
            print("Connection is not open")

    def close(self) -> None:
        if self.state == State.closed:
            print("Connection is not open")


connection = Connection()
connection.send_message("Hello")
connection.connect()
connection.send_message("Hello")
connection.connect()
connection.send_message(None)
connection.close()
